### PR TITLE
Auto-download emulators through emulators:start

### PR DESCRIPTION
**Pull request by samtstern**
### Description

Changes:
  * When using `emulators:start` the Java emulators will download if necessary.
  * Downloads now show a visual progress bar.
	 
### Scenarios Tested

**Before**
```bash
$ npx firebase-tools@7.0.0 emulators:start
i  Starting emulators: ["functions","firestore","database"]
✔  functions: Using node@8 from host.
✔  functions: Emulator started at http://localhost:5001
⚠  Setup required, please run: firebase setup:emulators:firestore
i  Shutting down emulators.
i  Stopping functions emulator

Error: An unexpected error has occurred.
```

**After**
```bash
$ npx firebase emulators:start
i  Starting emulators: ["functions","firestore","database"]
✔  functions: Using node@8 from host.
✔  functions: Emulator started at http://localhost:5001
i  firestore: downloading emulator...
Progress: ====================================================================> (100%)
i  firestore: Logging to firestore-debug.log
✔  firestore: Emulator started at http://localhost:8080
i  firestore: For testing set FIRESTORE_EMULATOR_HOST=localhost:8080
i  database: downloading emulator...
Progress: ====================================================================> (100%)
i  database: Logging to database-debug.log
✔  database: Emulator started at http://localhost:9000
```

```bash
$ npx firebase serve --only database
⚠  Setup required, please run: firebase setup:emulators:database

Error: An unexpected error has occurred.
```

### Sample Commands

N/A

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -3244,18 +3228,15 @@
...
         "code-point-at": {
           "version": "1.1.0",
-          "bundled": true,
-          "optional": true
+          "bundled": true
         },
         "concat-map": {
           "version": "0.0.1",
-          "bundled": true,
-          "optional": true
```
Comment (ryanpbrewster) [inline package-lock.json]: any idea why these changes keep bouncing around? different versions of `npm`?

Comment (samtstern) [reply package-lock.json]: @abeisgoat explain JavaScript to me plz

File: src/serve/javaEmulators.ts [hunk 2]
```diff
@@ -167,12 +178,18 @@ export async function stop(targetName: JavaEmulators): Promise<void> {
...
-    utils.logWarning("Setup required, please run: firebase setup:emulators:" + emulator.name);
-    return Promise.reject("emulator not found");
+
+  const hasEmulator = fs.existsSync(emulator.localPath);
+  if (!hasEmulator) {
+    if (args.auto_download) {
+      await downloadEmulator(targetName);
+    } else {
+      utils.logWarning("Setup required, please run: firebase setup:emulators:" + emulator.name);
+      return Promise.reject("emulator not found");
```
Comment (bkendall) [inline src/serve/javaEmulators.ts]: Can this simply `throw` rather than returning a rejected `Promise`?

File: package-lock.json (hunk 1 above)
Comment (yuchenshi) [reply package-lock.json]: tldr: Please use npm >= 6.9.0. All reviewers should ask for npm >= 6.9.0 if they see diffs like this. This is documented at https://github.com/firebase/firebase-tools/blob/master/.github/CONTRIBUTING.md#development-setup

Comment (samtstern) [reply package-lock.json]: Thanks Yuchen!

File: src/serve/javaEmulators.ts [hunk 3]
```diff
@@ -73,6 +85,11 @@ function _getCommand(emulator: JavaEmulators, args: { [s: string]: any }): JavaE
 
   const cmdLineArgs = baseCmd.args.slice();
   Object.keys(args).forEach((key) => {
+    if (baseCmd.optionalArgs.indexOf(key) < 0) {
```
Comment (yuchenshi) [inline src/serve/javaEmulators.ts]: Why this change? I'd suggest that we just error when we see an unsupported flag.

Comment (samtstern) [reply src/serve/javaEmulators.ts]: This is the first time we want to add a flag to the in-memory `FooEmulatorArgs` that we don't want to pass directly to the JAR (`auto_download`).  So this filters those out.
